homeassistant/components/binary_sensor/w800rf32.py

```diff
@@ -0,0 +1,170 @@
+"""
+Support for w800rf32 binary sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.w800rf32/
+
+```yaml
+# Example configuration.yaml entry
+
+binary_sensor:
+  - platform: w800rf32
+    devices:
+      c1:
+        name: motion_hall
+        off_delay: 5
+        device_class: Motion
+      c2:
+        name: motion_kitchen
+        device_class: Motion
+        fire_event: True
+
+```
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import w800rf32
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA, PLATFORM_SCHEMA, BinarySensorDevice)
+from homeassistant.components.w800rf32 import (CONF_FIRE_EVENT, CONF_OFF_DELAY)
+from homeassistant.const import (CONF_DEVICE_CLASS, CONF_NAME, ATTR_NAME, CONF_DEVICES)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import event as evt
+from homeassistant.util import dt as dt_util
+
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['w800rf32']
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICES, default={}): {
+        cv.string: vol.Schema({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_FIRE_EVENT, default=False): cv.boolean,
+            vol.Optional(CONF_OFF_DELAY):
+            vol.Any(cv.time_period, cv.positive_timedelta)
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Binary Sensor platform to w800rf32."""
+    import W800rf32 as w800rf32mod
+    sensors = []
+
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary) --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        if device_id in w800rf32.W800_DEVICES:
+            continue
+
+        _LOGGER.debug("Add %s w800rf32.binary_sensor (class %s)",
+                      entity[ATTR_NAME], entity.get(CONF_DEVICE_CLASS))
+
+        event = None  # None until an event happens
+        device = W800rf32BinarySensor(
+            event, entity.get(CONF_NAME), entity.get(CONF_DEVICE_CLASS),
+            entity[CONF_FIRE_EVENT], entity.get(CONF_OFF_DELAY))
+
+        sensors.append(device)
+        w800rf32.W800_DEVICES[device_id] = device
```

Comment:
We're not allowed to store entities in a shared dictionary or access entities directly from outside their platforms. Please use our dispatch helper to signal entities from component to platform. It's ok to store a reference to the entity in `hass.data` eg the entity_id.

See the tuya component for an example of how to use our dispatch helper.

Reply:
So can I keep a list inside the component, if it's only ever used inside the component and not the platform?

Reply:
No, entities are a platform concern.

Reply:
Right so I'm keeping a list of entities in hass.data[DOMAIN]['entities'], that should be fine correct?

Oh and by the way..thank you for your help.  

I would really like to get this perfect and then update some of the docs so people won't have these types of questions and problems.

Reply:
Why do we need a list of the entities? We should use the dispatch helper to signal the correct entity.

Reply:
ok 'I'll try to figure that out. I looked at it but of course there is no docs other than code.

Reply:
Yes, the docs on this are minimal unfortunately.

https://developers.home-assistant.io/docs/en/creating_component_code_review.html#3-component-platform-communication

Reply:
Maybe there is confusion, at this point I'm only talking about storing data, which those docs say store in hass.data[DOMAIN] or hass.data[DATA_XY], which is what I'm doing now.

Reply:
It's ok to store data, but not the entity instances, to be accessed directly from the component.

Reply:
So the entity instances are only ever stored in 'switches', 'binary_sensors' etc?

Reply:
The entities are stored in the entity platform helper after calling `add_entities`.

Sometimes its necessary to store the entities also directly in the platform, eg when registering a service in the platform. But that's not the case here.

Reply:
No I was thinking data structures not methods.
Is it supposed to look like this?
/
/hass/data/switch/_platforms/w800rf32/entities
                                                                  ^
and then all your entities under this        ---

Reply:
Sorry for asking so many questions but I'm really trying to understand the entire picture here so I can eventually add some more complete documentation.

Reply:
Ok so I completely confused now, If I can't store a list of my entity instances how do I dig it out of hass so I can call update_status() from my component?

Reply:
By using the dispatch helper, we don't need direct access to the entities.

Reply:
Just to add, even using dispatch helper I still need to know which entity I'm dealing with in dispatch_connect and send, if I'm reading that correctly.

Reply:
I guess what I'm saying is somehow or other I need to connect a hardware address, ex: 'c1' with and instance in HA so incoming data can operate on the correct entity.  This is simple when I keep a list of then inside my component.

Reply:
Here's how we connect a callback to an entity/device specific signal in mysensors component:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mysensors/device.py#L104-L109

Here's how the elkm1 alarm control panel signals a list of entity_ids and passes arguments to the connected function.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alarm_control_panel/elkm1.py#L50-L53

It looks like we can use the `device_id` as an identifier. Store this with each entity, and use that to create device specific signals. When an event comes in, extract the device_id, create the signal string and send the signal with the dispatcher send function.

Sounds good?

Reply:
Many thanks.

As to your last paragraph, that's pretty much what I was doing, other components such as rxfrx do the same, except we keep our data local to the component and platform, which in my mind is where it should be, it has no other meaning to HA.  It's like storing local data about a hardware device.

It's much easier to follow and I don't really see any benefit to using the dispatcher in these particular cases.  Do you happen to know the purpose of forcing the use of it, out of curiosity?

Reply:
Entities are a detail of the platforms that create them. We don't access them directly from outside. Limiting interaction points between different parts of home assistant makes it easier to reason about the logic and predict behavior.